driver/parallel.go

```diff
@@ -0,0 +1,222 @@
+package driver
+
+import (
+	"sync"
+
+	"github.com/brimsec/zq/ast"
+	"github.com/brimsec/zq/expr"
+	"github.com/brimsec/zq/filter"
+	"github.com/brimsec/zq/proc"
+	"github.com/brimsec/zq/scanner"
+	"github.com/brimsec/zq/zbuf"
+	"github.com/brimsec/zq/zng"
+)
+
+type parallelHead struct {
+	proc.Base
+	once sync.Once
+	pg   *parallelGroup
+
+	mu sync.Mutex // protects below
+	sc ScannerCloser
+}
+
+func (ph *parallelHead) closeOnDone() {
+	select {
+	case <-ph.Context.Done():
+		ph.mu.Lock()
+		defer ph.mu.Unlock()
+		if ph.sc != nil {
+			ph.sc.Close()
+			ph.sc = nil
+		}
+	}
```

Comment:
Nits: No need for select or defer.
```suggestion
	<-ph.Context.Done()
	ph.mu.Lock()
	if ph.sc != nil {
		ph.sc.Close()
		ph.sc = nil
	}
	ph.mu.Unlock()
```